### ISSUE TITLE
Feature request: `useShouldUpdate` hook

### ISSUE THREAD

**Issue by josepot** [facebook/react#14946]
**Do you want to request a *feature* or report a *bug*?**
Request a feature

It is currently impossible to let React know that the output of a **functional component** is not affected by the current change. I think that it would be great to have a hook for enabling this capability. In other words: having the SCU functionality in functional components through a hook.

I'm aware of the existence of `React.memo` and I know that it is possible to accomplish the same thing by splitting the logic in 2 different functional components... as long as we enhance the "base" component with `React.memo`. However, I still think that the hook that I'm suggesting would be a pretty nice addition.

I guess that it's a good idea to show an example of a real case where this hook would be helpful. So, here we go:

I don't normally use the official react-redux bindings. Instead, I have my own version of the `connect` HOC which has a more limited (and slightly different) API, that better suits my needs. This makes my version a bit more performant and a lot lighter. Regardless of whether it is a good idea not to use the official react-redux bindings, the following example illustrates the benefit of having the hook that I'm suggesting.

This is the current implementation of my custom "connect" HOC:

```js
const emptyObj = {};                                                           
const alwaysEmpty = () => emptyObj;                                            
                                                                               
export default (fromStateProps_, fromActionProps = emptyObj, mapper) => {      
  const dependsOnProps = fromStateProps_ && fromStateProps_.length !== 1;      
  const fromStateProps = !fromStateProps_ ? alwaysEmpty : fromStateProps_;     
                                                                               
  return BaseComponent => {                                                    
    const MBaseComponent = memo(BaseComponent);                                
                                                                               
    return props => {                                                          
      const dependantProps = dependsOnProps ? props : emptyObj;                
                                                                               
      const {state, dispatch} = useContext(context);                           
      const stateProps = useMemo(() => fromStateProps(state, dependantProps), [
        state,                                                                 
        dependantProps,                                                        
      ]);                                                                      
      const actionProps = useMemo(() => {                                      
        const res = {};                                                        
        Object.entries(fromActionProps).forEach(([name, aCreator]) => {        
          res[name] = (...args) => dispatch(aCreator(...args));                
        });                                                                    
        return res;                                                            
      }, [dispatch]);                                                          
                                                                               
      return useMemo(() => {                                                   
        const finalProps = mapper                                              
          ? mapper(stateProps, actionProps, props)                             
          : {...props, ...stateProps, ...actionProps};                         
        return <MBaseComponent {...finalProps} />;                             
      }, [actionProps, props, stateProps]);                                    
    };                                                                         
  };                                                                           
};                                                                             
```

It works great (for me). However, I would much rather to be able to implement it like this:

```js
const emptyObj = {};                                                           
const alwaysEmpty = () => emptyObj;                                            
                                                                               
export default (fromStateProps_, fromActionProps = emptyObj, mapper) => {      
  const dependsOnProps = fromStateProps_ && fromStateProps_.length !== 1;      
  const fromStateProps = !fromStateProps_ ? alwaysEmpty : fromStateProps_;     
                                                                               
  return BaseComponent => {                                                    
    const componentFn = BaseComponent.prototype                                
      ? p => <BaseComponent {...p} />                                          
      : p => BaseComponent(p);                                                 
                                                                               
    return props => {                                                          
      const dependantProps = dependsOnProps ? props : emptyObj;                
                                                                               
      const {state, dispatch} = useContext(context);                           
      const stateProps = useMemo(() => fromStateProps(state, dependantProps), [
        state,                                                                 
        dependantProps,                                                        
      ]);                                                                      
      const actionProps = useMemo(() => {                                      
        const res = {};                                                        
        Object.entries(fromActionProps).forEach(([name, aCreator]) => {        
          res[name] = (...args) => dispatch(aCreator(...args));                
        });                                                                    
        return res;                                                            
      }, [dispatch]);                                                          
                                                                               
      useShouldUpdate([actionProps, props, stateProps]);                       
                                                                               
      const finalProps = mapper                                                
        ? mapper(stateProps, actionProps, props)                               
        : {...props, ...stateProps, ...actionProps};                           
                                                                               
      return componentFn(finalProps);                                          
    };                                                                         
  };                                                                           
};                                                                             
```

How do you feel about this suggestion? Too weird? 😄 

Thanks,

Josep

**Comment by aweary** [comment]
> I think that it would be great to have a hook for enabling this capability. In other words: having the SCU functionality in functional components through a hook.

You can accomplish something very similar to this using the `useMemo` hook. The [Hooks FAQ docs](https://reactjs.org/docs/hooks-faq.html#how-to-memoize-calculations) mention this:

> Conveniently, useMemo also lets you skip an expensive re-render of a child:

```jsx
function Parent({ a, b }) {
  // Only re-rendered if `a` changes:
  const child1 = useMemo(() => <Child1 a={a} />, [a]);
  // Only re-rendered if `b` changes:
  const child2 = useMemo(() => <Child2 b={b} />, [b]);
  // ...
}
```

**Comment by josepot** [comment]
Dah! You are absolutely right @aweary ! Thanks a lot for your quick response...

**Comment by josepot** [comment]
Hi @aweary ,

I'm reopening this issue because I'm afraid that returning the same reference from a render function (using `useMemo`) like you were suggesting is not equivalent to returning false from the `shouldComponentUpdate` lifecycle.

[The following codesandbox shows that](https://codesandbox.io/s/zr8zwlp03).

I'm not sure whether this is a bug with React (perhaps it is), but I thought that returning `false` from SCU was "something more" than just returning the same reference from the `render` function. I thought that since returning false from the SCU completely prevents the render from happening, that was a more "powerful hint" than just returning the same reference from the `render` function.

So, now I'm wondering whether what's happening in the codesandbox is a bug with React, because if it is not, then perhaps it does make sense to add the hook that I'm suggesting.

**Comment by gaearon** [comment]
I wrote about why a Hook like this wouldn't make sense:

https://overreacted.io/why-isnt-x-a-hook/#not-a-hook-usebailout

Hope it helps.

As for your latest sandbox, you can see that it works if you remove React Router: https://codesandbox.io/s/rwl6mvpjx4. So the issue is due to you using legacy context API (through React router) which disables the memoization optimization. Avoid legacy context.

**Comment by josepot** [comment]
Hi @gaearon ,

Thanks a lot for your quick response and for sharing that link, it does help a lot.

I was under the impression that `react-router` 4 was still using the legacy context API... I actually looked at the code of their official repo and I saw that they were using the new context API (I should have looked at the code inside the `node_modules` folder instead :see_no_evil: ). The following sandbox uses their latest beta version: https://codesandbox.io/s/w6z8xr3x0l and the problem goes away.

So, yep, that explains, thanks a lot!
